로그인 페이지 및 회원가입 양식 페이지 구현

riverandeye opened the pull request:
## 작업 내용

1. 로그인 페이지와 회원가입 양식 페이지에 Grid를 적용하였습니다. 
2. 로그인 페이지에 AppBar를 추가하였고, AppBar를 페이지별로 커스터마이즈 할 수 있게끔 props로 값을 전달할 수 있게 설정하였습니다.
3. 회원가입 양식 페이지의 입력에 대한 오류 메세지를 각 입력 바 밑에 나타나도록 추가하였습니다.
4. 입력한 username을 가진 회원을 찾는 api를 구현하였습니다.
5. username 입력시 매번 api에 요청하지 않게끔 debounce 기법을 이용하여 구현하였습니다

## 실행 방법

http://localhost:3000/auth/signup 에서 확인해보기!

## 결과 첨부

![중복체크](https://user-images.githubusercontent.com/41494099/71027706-211e7a80-214f-11ea-8bfb-3318907f3dde.gif)
![요청여부](https://user-images.githubusercontent.com/41494099/71027707-224fa780-214f-11ea-8f81-fec9331ded57.gif)

riverandeye committed:
[client] | refactor(Login): Login 페이지에 Grid 적용

로그인 페이지에 그리드를 적용하였습니다.

riverandeye committed:
[client] | refactor(SignUp): 회원 가입 양식 페이지 Grid 적용

회원 가입 양식 페이지에 그리드를 적용하였습니다.

riverandeye committed:
[client] | feat(Login): 로그인 페이지 디자인 변경사항 반영

로그인 페이지 디자인 변경사항을 반영하였습니다.

riverandeye committed:
[client] | feat(AppBar): 앱바 props 추가

앱바의 기능을 확장하기 위하여 props를 추가하여 적용하였습니다.
backgroundColor : 앱바의 배경 색 지정
showSearchBar: 검색 바를 포함할지 말지를 지정
showButtons:  버튼들을 포함할지 말지를 지정

riverandeye committed:
[client] | feat(Login): 로그인 페이지에 앱바 추가

로그인 페이지에 앱바를 추가하였습니다.

riverandeye committed:
[client] | feat(SignUp): 회원가입 양식 오류 메세지 추가

회원가입 시 오류에 해당하는 메세지의 마크업을 추가하였습니다.

riverandeye committed:
Merge branch 'develop' of https://github.com/connect-foundation/2019-10 into feature/login_ui_20191204

riverandeye committed:
[client] | feat(SignUp): 회원가입 양식 입력 validation 및 메세지 추가

회원 가입 페이지에 올바르지 않은 값 입력시 입력 바 하단에 나타나는 경고 메세지를 추가하였고, 가입하기 기능이 동작하지 않게 구성하였습니다.

riverandeye committed:
[server] | feat(user): username을 통한 사용자 검색 기능 추가

username을 이용하여 사용자를 검색하는 기능이 추가되었습니다.

riverandeye committed:
[client] | feat(SignUp): 회원 가입 시 사용자 입력 validation 기능 추가

회원 가입시 사용자 입력에 대한 validation을 하는 기능을 추가하였습니다.

riverandeye committed:
[client] | feat(SignUp): userName 중복 확인 및 디바운스 적용

username 에 대하여 디바운스를 적용하여, 매 입력마다 상태가 변경되지만, 서버에 중복 체크를 매 입력마다 하지 않습니다.

riverandeye committed:
[server] | feat(user): username을 validate 하는 pipe 추가

username에 해당하는 사용자 검색 시 param으로 입력되는 username이 올바른 조건의 입력인지 검증하는 파이프를 추가하였습니다.

KimJye commented on packages/server/src/user/user.controller.ts:
```diff
@@ -41,11 +43,25 @@ export class UserController {
+    try {
+      return new UserResponseDto(user);
+    } catch (err) {
+      if (!user) {
```
현재 service에서 작성하신 findUserByName 함수에서, user를 찾지 못한 경우 404 에러를 발생하지 않나요?? 맞다면 이 부분을 수정해야할것같아요~

Atercatus replied:
아마 빈 값이 넘어 오는 걸루 알고있는데 한번 더 확인 해주세요~

riverandeye replied:
네 꼼꼼한 리뷰 감사해요~ 말씀하신대로 오류가 존재했는데, regex에 g 태그를 추가한 것 때문에 발생한 오류였습니다. 수정한 이후 검색한 username에 해당하는 사용자가 없는 경우에 빈 값이 반환됩니다!

Atercatus commented on packages/client/libs/validate.ts:
```diff
@@ -0,0 +1,38 @@
+  userName: string,
+  isDuplicated: boolean,
+): [boolean, string] => {
+  if (isDuplicated) {
```
type을 정의하는게 이해가 빠를 거 같아요

Atercatus commented on packages/client/libs/validate.ts:
```diff
@@ -0,0 +1,38 @@
+      `닉네임 '${userName}'을 사용중인 회원이 존재합니다. 다른 닉네임을 사용해주시길 바랍니다.`,
+    ];
+  }
+  if (!userName || userName.length === 0) {
```
두개의 조건이 겹치는 거 같아요

riverandeye replied:
반영하였습니다.

Atercatus commented on packages/client/libs/validate.ts:
```diff
@@ -0,0 +1,38 @@
+  if (!userName || userName.length === 0) {
+    return [true, ``];
+  }
+  if (userName.match(regex.userName)) {
```
조건들을 이해할 수 있게 함수로 묶으면 좋겠어요

riverandeye replied:
:) :+1:

Atercatus commented on packages/client/views/SignUp/hooks.ts:
```diff
@@ -0,0 +1,85 @@
+
+export const userFormReducer = (state, action) => {
+  switch (action.type) {
+    case 'reset': {
```
문자열은 변수로 빼면 좋을 거 같습니당.

riverandeye replied:
^^7

hyokualexkwon replied:
reducer type들은 별도의 types 와 같은 파일을 만들어서 관리하면 좋겠네요~

Atercatus commented on packages/client/views/SignUp/hooks.ts:
```diff
@@ -0,0 +1,85 @@
+import { responseStatus } from '../../response';
+import { endpoint } from '../../constants';
+
+export const initialUserState = {
```
명시된 타입이 있으면 좋을 거 같습니다.

Atercatus replied:
여기서 사용하는 state를 가리키는 타입이 있으면 좋겠네요

riverandeye replied:
반영하였씁니다~

Atercatus commented on packages/client/views/SignUp/hooks.ts:
```diff
@@ -0,0 +1,85 @@
+  }
+};
+
+export const useFormSubmit = (
```
매개변수의 개수를 줄이고 해당 class나 interface를 만들어주세요.
그 후에 destructing 해주세요

Atercatus commented on packages/client/views/SignUp/index.tsx:
```diff
@@ -1,110 +1,152 @@
 const SignUp: React.FunctionComponent = () => {
-  const [formData, setFormData] = useState(new FormData('', '', false));
-  const [isFetching, setIsFetching] = useState(false);
+  const [userForm, dispatchUserForm] = useReducer(
```
`userFormReducer` 와 `initialUserState` 를 반환하기보다 `useReducer`를 한 후 반환된 값을 export하는 게 좋을 거 같습니다.

Atercatus commented on packages/client/views/SignUp/index.tsx:
```diff
@@ -1,110 +1,152 @@
-    const { checked, name, value, type } = e.target;
-    const valueToUpdate = type === 'checkbox' ? checked : value;
+  useEffect(() => {
+    const shouldCheckDuplicated = () => {
```
함수명에 should는 왜 넣으신 건가요 어떤 의미를 가지나요?

Atercatus commented on packages/client/views/SignUp/index.tsx:
```diff
@@ -1,110 +1,152 @@
-    });
-  };
+    const checkUserNameDuplicate = async () => {
+      const user = await fetch(
```
fetch 대신에 라이브러리를 써야겠죠

Atercatus commented on packages/client/views/SignUp/index.tsx:
```diff
@@ -1,110 +1,152 @@
-      window.alert('이미 가입된 유저입니다.');
+    if (shouldCheckDuplicated()) {
+      checkUserNameDuplicate();
+    } else if (userForm.isUserNameDuplicated) {
```
else가 없는데 else if 가 존재하는 이유가 있을까욥

Atercatus commented on packages/client/views/SignUp/index.tsx:
```diff
@@ -1,110 +1,152 @@
-      router.push(endpoint.hotlist);
-    }
-  };
+  const isFormValid =
```
이름은 함수인데 변수인거 같습니다.

Atercatus commented on packages/client/views/SignUp/index.tsx:
```diff
@@ -1,110 +1,152 @@
+    userForm.isAgreed &&
+    !userForm.isUserNameDuplicated;
+
+  const { handleSubmit, checkSubmitAvailable } = useFormSubmit(
```
사실상 이런부분들을 hooks에 숨기면 자연스럽게 매개변수가 줄어들거 같네요. 안에서 해야할 동작을 밖에서 해버리니 같이 동작되어야할 것들이 따로 동작하게 되면서 내부동작인데 그 매개변수가 외부동작으로 부터 생성이 되니 그 값을 매개변수로 주게되는 현상인거 같습니다.
말이 명확하지 못한데 좀 더 상세한 부분은 저한테 말씀주시면 대면으로 말씀드리겠습니다,

Atercatus commented on packages/client/views/SignUp/index.tsx:
```diff
@@ -1,110 +1,152 @@
-  const checkSubmitAvailable = () => {
-    return !isFetching && checkFormVerified();
-  };
+  const handleDescription = e =>
```
이런 부분들도 hooks 내부에 위치시키면 깔끔할 거 같습니다.

riverandeye replied:
확인했습니다~

Atercatus commented on packages/server/src/common/regexes.ts:
```diff
@@ -1 +1,2 @@
 export const NATURAL_NUMBER_REGEX = /^[1-9]\d*$/;
+export const USER_NAME_REGEX = /^[\s0-9a-zA-Z\u3131-\uD79D]{1,30}$/giu;
```
`g` 를 지워주세욥!

Atercatus commented on packages/server/src/user/pipe/user-name-param-pipe.ts:
```diff
@@ -0,0 +1,12 @@
+@Injectable()
+export class UserNameParamPipe implements PipeTransform {
+  public async transform(userName: string): Promise<string> {
+    if (!USER_NAME_REGEX.test(userName)) {
```
함수로 분리해서 무슨 의민지 알려주면 좋겠죠

riverandeye replied:
반영하였습니다.

riverandeye committed:
[client] | fix(regex): regex의 g modifier 삭제

regex의 g modifier를 사용시 state를 갖게 되어, 예상치 못한 동작을 하게 됩니다. 그러한 이유로 g 태그를 모두 제거합니다.

riverandeye committed:
[client] | refactor(SignUp): validate 반환형 설정

validateResult class를 만들어 리턴 값을 명시하였습니다.

hyokualexkwon commented on packages/client/components/AppBar/index.tsx:
```diff
@@ -8,12 +8,13 @@ import * as S from './styles';
+export const AppBar: React.FunctionComponent<AppBarProps> = ({
+  backgroundColor = '#383d3f',
+  showSearchBar = true,
+  showButtons = true,
```
show라는 동사를 제외하고  searchBar, buttons 라고 표현하는건 어떨까요?

riverandeye replied:
반영하였습니다 👍

hyokualexkwon commented on packages/client/views/SignUp/index.tsx:
```diff
@@ -1,110 +1,152 @@
+                  id="username"
+                  name="username"
+                  maxLength={signUpFormDataMaxLength.username}
+                  onChange={handleUserName}
```
본 input에 value를 넘기지 않고  onChange만 작성해도 상관없나요? 리액트가 warning 발생시키지는 않나요?

riverandeye replied:
네 에러가 없습니다 👍

hyokualexkwon commented on packages/client/views/Login/index.tsx:
```diff
@@ -1,29 +1,43 @@
-    </S.Login>
+    <>
+      <AppBar showSearchBar={false} showButtons={false} />
+      <S.Login>
```
AppBar를 S.Login하위에 위치시키고 fragment를 삭제해도 좋을것 같습니다. 수정해보니까 ui는 문제없이 나타나네요.

riverandeye replied:
👍 반영하였습니다.

hyokualexkwon commented:
로그인화면에서 앱바의 색상을 배경화면과 일치시켜주세요.
![image](https://user-images.githubusercontent.com/17868599/70689181-e18f0300-1cf6-11ea-8a94-f3c4c3dde66e.png)

hyokualexkwon commented on packages/client/views/SignUp/styles.tsx:
```diff
@@ -27,16 +28,17 @@ export const HeadMessage = styled.div`
   margin-bottom: 5.5rem;
```
요기 위의 font-weight을 800으로 바꿔주세요

riverandeye replied:
반영하였습니다 :)

hyokualexkwon commented on packages/client/views/SignUp/styles.tsx:
```diff
@@ -62,15 +64,9 @@ export const Item = styled.div`
     border-radius: 0.5rem;
-    margin-bottom: 3rem;
     color: white;
     padding: 1rem;
```
input에 font-size를 명시해 주세요. 16px입니다.

riverandeye replied:
👍 :)

hyokualexkwon commented on packages/client/views/SignUp/styles.tsx:
```diff
@@ -89,20 +86,22 @@ export const Item = styled.div`
+    display: ${props => (props.valid ? 'none' : 'inline')};
+    margin-top: 1rem;
+    font-size: 1.4rem;
+    font-weight: bold;
```
700 🏓

riverandeye replied:
👍

hyokualexkwon commented on packages/client/views/SignUp/styles.tsx:
```diff
@@ -122,6 +121,9 @@ export const Label = styled.div`
     font-weight: bold;
```
700 🏓

riverandeye replied:
👍

hyokualexkwon commented on packages/client/views/SignUp/styles.tsx:
```diff
@@ -140,15 +141,9 @@ export const SubmitButton = styled.div`
     font-weight: bold;
```
700 🏓

riverandeye replied:
👍

hyokualexkwon reviewed (changes requested):
고생 많으셨네요!

riverandeye committed:
[server] | fix(regex): Regex의 g modifier 제거

state를 저장하여 예상치 못한 결과를 반환하게 하는 g modifier를 제거합니다.

riverandeye committed:
[client] | refactor(AppBar): Props 이름 변경

riverandeye committed:
[client] | refactor(Login): 컴포넌트 위치 변경 및 Fragment 제거

불필요한 Fragment를 제거합니다.

riverandeye committed:
[client] | refactor(SignUp): 글씨체 굵기 값 모두 변수로 적용

riverandeye committed:
Merge branch 'develop' of 2019-10 into feature/login_ui_20191204

Atercatus commented on packages/client/constants.ts:
```diff
@@ -89,7 +89,16 @@ export const fontWeight = {
+  description: 1500,
+};
+
+export const regex = {
```
이후에 작성하는 것들은 모두 대문자로 하기로 했쬬

Atercatus commented on packages/client/libs/validate.ts:
```diff
@@ -0,0 +1,39 @@
+import { regex, signUpFormDataMaxLength } from '../constants';
```
이 파일 전체는 lib에 위치하면 안될거 같네요

Atercatus replied:
리뷰 반영이 되었는지 꼭 다시 한번 더 확인하고 PR 요청해주세요

riverandeye replied:
죄송..ㅠㅠ

riverandeye committed:
Merge branch 'develop' of https://github.com/connect-foundation/2019-10 into feature/login_ui_20191204

riverandeye committed:
[client] | refactor(env): API_SERVER_HOST -> API_SERVER_URL 로 변경

riverandeye committed:
[client] | refactor(constant): 상수 이름 변경

riverandeye committed:
[client] | refactor(SignUp): 글씨 굵기 오류 수정

riverandeye committed:
[client] | feat(lib): ValidationState 정의

입력이 validate 된 결과 및 메세지를 저장하는 validationState를 작성하였습니다.

riverandeye committed:
[client] | feat(SignUp): 사용자 입력 정보 , validaton 결과 각각에 대한 모델 정의

사용자가 입력한 정보는 userFormState로, 그 값을 validation 한 결과는 UserFormValidationState로 정의됩니다.

riverandeye committed:
[client] | fix(SignUp): CSS 수정사항 반영

riverandeye committed:
[client] | feat(SignUp): 회원 가입 양식 dto 및 factory 정의

sign-up-form에 해당하는 Dto와 입력된 값을 이용하여 sign-up-form을 생성해주는 DtoFactory

riverandeye committed:
[client] | feat(SignUp): 회원 가입 Action을 생성하는 함수 작성

riverandeye committed:
[client] | feat(SignUp): Validate 함수 및 ValidationState를 생성하는 Factory 작성

riverandeye committed:
[client] | feat(SignUp): SignUp 컴포넌트 내에 사용될 훅 정의

riverandeye committed:
[client] | feat(SignUp): 회원 가입 페이지 마크업 변경 및 작성 및 훅 적용

riverandeye committed:
[client] | refactor(SignUp): 초기 validation state 변경 및 불필요한 함수 제거

user의 validation 초기 값을 false로 지정해야 사용자가 회원가입 양식을 제출할 수 없게 됩니다.

riverandeye committed:
[client] | feat(SignUp): username으로 사용자 정보를 조회하는 action 만드는 함수 생성

riverandeye committed:
[client] | feat(SignUp): username 중복을 validate 하는 함수 작성

username과 query를 입력하면, query에 입력된 username이 있으면 validate 실패 상태와 메세지를, 아니면 validate 성공 상태를 반환합니다.

riverandeye committed:
[client] | feat(SignUp): 입력의 중복 여부를 저장하는 State

riverandeye committed:
[client] | feat(SignUp): 회원 가입 훅에 username 중복 체크 기능 추가

riverandeye committed:
[client] | feat(SignUp): username 중복 메세지 보여주는 기능 추가

입력한 username이 중복되는 경우 에러 메세지를 보여주는 기능을 추가하였습니다.

riverandeye committed:
[client] | refactor(SignUp): 매직 넘버 제거 및 변수명 지정

Atercatus commented on packages/client/constants.ts:
```diff
@@ -38,6 +38,7 @@ export const endpoint = {
   login: '/auth/login',
   users: '/users',
   profileEdit: '/profile/edit',
+  verifyUsername: '/users/verify',
```
verifyUsername 이라는 이름보다는 verifyUserInfo 가 낫지 않을까요. 쿼리스트링으로 username= 이렇게 주시는거 같은데

riverandeye replied:
좋은 제안 반영하겠습니다.

Atercatus commented on packages/client/constants.ts:
```diff
@@ -112,3 +117,14 @@ export const TITLE = 'title';
+  INVALID_CHARACTER: `특수 문자는 포함할 수 없습니다.`,
+};
+
+export const SERVER_ERROR_MESSAGE = `알 수 없는 오류가 발생했습니다. 새로고침하고 다시 시도해주세요.`;
```
이런 메시지는 작성하시면 안됩니다. 알 수 없는 오류라는걸 사용자에게 알리면 망신이겠죠?

riverandeye replied:
반영하였습니다~

Atercatus commented on packages/client/views/SignUp/action/make-get-user.ts:
```diff
@@ -0,0 +1,8 @@
+
+import { endpoint } from '../../../constants';
+
+export const makeGetUserAction: Action = (username: string) => ({
```
Query로 하기로 했었죠. 관훈님은 항상 정한 컨벤션을 못지키는 버릇이 있는거 같습니다.  어딘가에 기록을 해두고 PR 날리시기전에 점검하고 리뷰요청해주세요

riverandeye replied:
반영하였습니다~

Atercatus commented on packages/client/views/SignUp/dto/sign-up-form-dto.ts:
```diff
@@ -0,0 +1,4 @@
+export interface SignUpFormDTO {
+  username: string;
+  description: string;
```
DTO에 약관 ㄴ동의 정보가 왜 사라져있죠? 클라이언트에서 이런것을 누락시키면 안됩니다.

Atercatus replied:
동의 했는지 서버에서 기록해야 될 부분이지 동의안했으면 요청 안해야지라는 생각으로 클라이언트 딴에서 데이터를 누락시키면 안됩니다.

riverandeye replied:
반영했습니다~

Atercatus commented on packages/client/views/SignUp/hook/use-sign-up.ts:
```diff
@@ -0,0 +1,118 @@
+  );
+
+  const [userFormValidationState, setUserFormValidationState] = useState(
+    new UserFormValidationState(),
```
UserFormValidationState's' 겠죠

Atercatus commented on packages/client/views/SignUp/model/duplication-check-state.ts:
```diff
@@ -0,0 +1,9 @@
+import { ValidationState } from '../../../libs/validation-state/validation-state';
+
+export class DuplicationValidationState {
```
분리했어야만했나요

Atercatus commented on packages/server/src/user/user.controller.ts:
```diff
@@ -45,6 +48,21 @@ export class UserController {
+  public async getUserWithName(
+    @Param('username', new UserNameParamPipe()) username: string,
+  ): Promise<UserResponseDto | object> {
+    const user = await this.userService.findUserByName(username);
```
이게 try 밖에있는데 catch는 왜할까요

riverandeye replied:
반영하였습니다~

Atercatus reviewed (changes requested):
많은 부분이 리펙토링이 필요한거 같습니만 일정상문제로 우선은 넘어가더라도 다른 문제들이 있어서 다시 리퀘스트 날립니다. 꼭 PR 전에 점검사항을 다시 확인해주세요

riverandeye committed:
[server] | refactor(user): 사용자 이름 검사 Pipe 함수 분리

가독성 좋게 validation 하는 과정을 분리하였습니다.

riverandeye committed:
[client] | refactor(libs): 상수명 변경 및 불필요한 상수 제거

riverandeye committed:
[client] | refactor(libs): 사용되지 않는 코드 제거

riverandeye committed:
[client] | refactor(SignUp): username에 해당하는 사용자 찾는 액션 만드는 함수 이름 변경

riverandeye committed:
[client] | fix(libs): 회원가입 요청에 생략된 isAgreed 정보 추가

riverandeye committed:
[client] | refactor(SignUp): 불필요한 Validation 메세지 제거

riverandeye committed:
[client] | refactor(SignUp): ValidationState 모델 이름 복수로 변경

riverandeye committed:
[client] | refactor(SignUp): DuplicatioCheckState 모델 이름 복수로 변경

riverandeye committed:
[server] | refactor(user): 사용자 검색 api 에러처리문 변경

service에서 발생하는 오류를 예방하기 위하여 try 안에 포함하였습니다.

riverandeye committed:
[client] | refactor(SignUp): hook과 index에 함수 및 변수명 변경 수렴

riverandeye committed:
[client] | refactor(SignUp): Title 폰트 extraBold 설정

riverandeye committed:
Merge branch 'develop' into feature/login_ui_20191204

Atercatus commented on packages/server/src/user/user.controller.ts:
```diff
@@ -52,13 +52,13 @@ export class UserController {
       if (!user) {
         return {};
       }
+      return new UserResponseDto(user);
```
```javascript
if(user){
  return false;
}

return true;
```

Atercatus replied:
이런 방식으로 하면 좋을 거 같습니다.

Atercatus commented on packages/server/src/user/user.controller.ts:
```diff
@@ -52,13 +52,13 @@ export class UserController {
   public async getUserWithName(
```
네이밍이 좀 더 개선사항이 보입니다.

riverandeye committed:
[server] | feat(user): 중복 체크 결과를 반환하는 model 정의

riverandeye committed:
[server] | feat(user): username 중복 검증 api 함수명 변경 및 반환형 정의

riverandeye committed:
[client] | feat(SignUp): URL로 직접 접근할 수 없게끔 설정 및 리팩토링

URL로 직접 접근할 수 없게끔 getInitialProps에서 redirect 하였습니다.
validateUsernameDuplicate 함수 내에서 변수명을 적절하게 수정하였습니다.

Atercatus commented on packages/server/src/user/model/duplicate-check-result.ts:
```diff
@@ -0,0 +1,7 @@
+    this.value = value;
+  }
+
+  public readonly value: boolean;
```
`value` is too bad.

riverandeye committed:
[server] | refactor(user): 중복 체크 api 반환형 model의 변수명 올바르게 변경

Atercatus reviewed (approved):
이후 리펙토링 과정이 필요한거 같습니다.
